File: backend/apps/core/validation.py
from datetime import date, datetime
from decimal import Decimal

from django.utils import timezone


def is_blank(value):
    if value is None:
        return True
    if isinstance(value, str):
        return value.strip() == ''
    return False


def get_value(attrs, instance, key):
    if key in attrs:
        return attrs.get(key)
    if instance is None:
        return None
    return getattr(instance, key, None)


def normalize_datetime(value):
    if value is None:
        return None

    if isinstance(value, datetime):
        if timezone.is_naive(value):
            return timezone.make_aware(value, timezone=timezone.utc)
        return value.astimezone(timezone.utc)

    return value


def normalize_for_task_range(value):
    value = normalize_datetime(value)
    if isinstance(value, datetime):
        return timezone.localtime(value) if timezone.is_aware(value) else value
    return value


def date_only(value):
    value = normalize_for_task_range(value)
    if isinstance(value, datetime):
        return value.date()
    if isinstance(value, date):
        return value
    return None


def hours_until_end_of_day(current_time=None):
    current_time = normalize_for_task_range(current_time or timezone.now())
    if not isinstance(current_time, datetime):
        return Decimal('0')

    end_of_day = current_time.replace(hour=23, minute=59, second=59, microsecond=999999)
    remaining_seconds = max((end_of_day - current_time).total_seconds(), 0)
    return Decimal(str(remaining_seconds)) / Decimal('3600')


def task_hours_limit(start_date, end_date, current_time=None):
    start_date = normalize_for_task_range(start_date)
    end_date = normalize_for_task_range(end_date)

    if not start_date or not end_date:
        return None

    start_day = date_only(start_date)
    end_day = date_only(end_date)
    if not start_day or not end_day:
        return None

    if start_day == end_day:
        return hours_until_end_of_day(current_time=current_time)

    duration_days = (end_day - start_day).days + 1
    if duration_days <= 0:
        return Decimal('0')
    return Decimal(str(duration_days * 24))


def required_field_errors(attrs, instance, fields, *, partial=False):
    if partial:
        return {}

    errors = {}
    for key, label in fields:
        value = get_value(attrs, instance, key)
        if is_blank(value):
            errors[key] = f'{label} is required'

    if errors:
        errors['error'] = 'All fields are required'
    return errors


def date_order_error(start_date, end_date):
    start_date = normalize_datetime(start_date)
    end_date = normalize_datetime(end_date)
    if start_date and end_date and end_date < start_date:
        return 'End date cannot be before start date'
    return None


def estimated_hours_errors(start_date, end_date, estimated_hours):
    if estimated_hours is None:
        return None

    try:
        hours = Decimal(str(estimated_hours))
    except Exception:
        return 'Estimated hours must be a positive number'

    if hours <= 0:
        return 'Estimated hours must be a positive number'

    limit_hours = task_hours_limit(start_date, end_date)
    if limit_hours is None:
        return None

    if hours > limit_hours:
        if date_only(start_date) == date_only(end_date):
            return 'Estimated hours cannot exceed remaining time today'
        return 'Estimated hours cannot exceed task duration'

    return None


def is_within_range(child_start, child_end, parent_start, parent_end):
    child_start = normalize_datetime(child_start)
    child_end = normalize_datetime(child_end)
    parent_start = normalize_datetime(parent_start)
    parent_end = normalize_datetime(parent_end)

    if not parent_start or not parent_end or not child_start or not child_end:
        return True
    return child_start >= parent_start and child_end <= parent_end

File: frontend/src/utils/validation.js
const DATE_RE = /^\d{4}-\d{2}-\d{2}$/

const FIELD_LABELS = {
  name: 'Title',
  title: 'Title',
  description: 'Description',
  start_date: 'Start date',
  end_date: 'End date',
  due_date: 'End date',
  priority: 'Priority',
  assignee_id: 'Assignee',
  category_id: 'Category',
  estimated_hours: 'Estimated hours',
  email: 'Email',
}

const isEmpty = (value) => value === null || value === undefined || String(value).trim() === ''

const toLocalDate = (value) => {
  if (!value || !DATE_RE.test(String(value))) return null
  const date = new Date(`${value}T00:00:00`)
  return Number.isNaN(date.getTime()) ? null : date
}

const toDate = (value) => {
  return toLocalDate(value)
}

const addRequiredErrors = (data, requiredFields, errors) => {
  requiredFields.forEach((field) => {
    if (isEmpty(data[field])) {
      const label = FIELD_LABELS[field] || field
      errors[field] = `${label} is required`
    }
  })
}

export const calculateTaskHoursLimit = (data, referenceDate = new Date()) => {
  const start = toLocalDate(data.start_date)
  const end = toLocalDate(data.due_date || data.end_date)

  if (!start || !end) {
    return { limitHours: null, isSameDay: false, remainingHours: null, isValidTimeSelection: true }
  }

  const startDay = start.toDateString()
  const endDay = end.toDateString()

  if (startDay === endDay) {
    const now = referenceDate instanceof Date ? referenceDate : new Date(referenceDate)
    const endOfDay = new Date(now)
    endOfDay.setHours(23, 59, 59, 999)
    const remainingMs = endOfDay.getTime() - now.getTime()
    const remainingHours = Math.max(remainingMs / 3600000, 0)
    return {
      limitHours: remainingHours,
      isSameDay: true,
      remainingHours,
      displayHours: Number(remainingHours.toFixed(2)),
      isValidTimeSelection: remainingHours > 0,
    }
  }

  const durationDays = Math.max(Math.floor((end.getTime() - start.getTime()) / 86400000) + 1, 0)
  return {
    limitHours: durationDays * 24,
    isSameDay: false,
    remainingHours: null,
    displayHours: durationDays * 24,
    isValidTimeSelection: true,
  }
}

const addCommonDateAndHoursErrors = (data, errors, referenceDate = new Date()) => {
  const start = toDate(data.start_date)
  const end = toDate(data.due_date || data.end_date)

  if (data.start_date && !start) {
    errors.start_date = 'Date format is invalid'
  }

  const endField = data.due_date !== undefined ? 'due_date' : 'end_date'
  if ((data.due_date || data.end_date) && !end) {
    errors[endField] = 'Date format is invalid'
  }

  if (start && end && end < start) {
    errors[endField] = 'End date cannot be before start date'
  }

  const hours = Number(data.estimated_hours)
  if (data.estimated_hours !== undefined && data.estimated_hours !== '' && !Number.isFinite(hours)) {
    errors.estimated_hours = 'Estimated hours must be a positive number'
  }

  if (Number.isFinite(hours) && hours <= 0) {
    errors.estimated_hours = 'Estimated hours must be a positive number'
  }

  const { limitHours, isSameDay, isValidTimeSelection } = calculateTaskHoursLimit(data, referenceDate)
  if (Number.isFinite(hours) && limitHours !== null) {
    if (!isValidTimeSelection && isSameDay) {
      errors.estimated_hours = 'Invalid time selection'
    } else if (hours > limitHours) {
      errors.estimated_hours = isSameDay
        ? 'Estimated hours cannot exceed remaining time today'
        : 'Estimated hours cannot exceed task duration'
    }
  }

  return { start, end }
}

export const validateProject = (data, referenceDate = new Date()) => {
  const errors = {}
  addRequiredErrors(data, ['name', 'description', 'start_date', 'end_date'], errors)

  const { start, end } = addCommonDateAndHoursErrors(
    { ...data, due_date: undefined, estimated_hours: undefined },
    errors,
    referenceDate
  )

  if (start && end && end < start) {
    errors.end_date = 'End date cannot be before start date'
  }

  return {
    isValid: Object.keys(errors).length === 0,
    errors,
    generalError: Object.keys(errors).length ? 'All fields are required' : '',
  }
}

export const validateTask = (data, { project, referenceDate = new Date() } = {}) => {
  const errors = {}
  addRequiredErrors(
    data,
    ['title', 'description', 'start_date', 'due_date', 'priority', 'assignee_id', 'category_id', 'estimated_hours'],
    errors
  )

  const { start, end } = addCommonDateAndHoursErrors(data, errors, referenceDate)

  if (project?.start_date && project?.end_date && start && end) {
    const projectStart = new Date(project.start_date)
    const projectEnd = new Date(project.end_date)
    if (start < projectStart || end > projectEnd) {
      errors.start_date = 'Task date must be within project date range'
      errors.due_date = 'Task date must be within project date range'
    }
  }

  return {
    isValid: Object.keys(errors).length === 0,
    errors,
    generalError: Object.keys(errors).length ? 'All fields are required' : '',
  }
}

export const validateSubtask = (data, { parent, referenceDate = new Date() } = {}) => {
  const result = validateTask(data, { project: null, referenceDate })

  if (!result.isValid) {
    return result
  }

  const start = toDate(data.start_date)
  const end = toDate(data.due_date)

  if (parent?.start_date && parent?.due_date && start && end) {
    const parentStart = new Date(parent.start_date)
    const parentEnd = new Date(parent.due_date)
    if (start < parentStart || end > parentEnd) {
      result.errors.start_date = 'Subtask must be within parent task date range'
      result.errors.due_date = 'Subtask must be within parent task date range'
      result.isValid = false
      result.generalError = 'All fields are required'
    }
  }

  return result
}

export const extractApiError = (error, fallback = 'Something went wrong') => {
  const data = error?.response?.data
  if (!data) return fallback

  if (typeof data.error === 'string' && data.error.trim()) {
    return data.error
  }

  const firstValue = Object.values(data)[0]
  if (Array.isArray(firstValue) && firstValue[0]) return String(firstValue[0])
  if (typeof firstValue === 'string') return firstValue

  return fallback
}

export const isValidEmail = (email) => /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(String(email || '').trim())
